diff --git a/sdk/tlibcxx/include/__config b/sdk/tlibcxx/include/__config
index 7835a963..dbe05270 100644
--- a/sdk/tlibcxx/include/__config
+++ b/sdk/tlibcxx/include/__config
@@ -411,6 +411,8 @@

 #if defined(_LIBCPP_COMPILER_CLANG)

+#include <__sgx>
+
 #if defined(_LIBCPP_ALTERNATE_STRING_LAYOUT)
 #  error _LIBCPP_ALTERNATE_STRING_LAYOUT is deprecated, please use _LIBCPP_ABI_ALTERNATE_STRING_LAYOUT instead
 #endif
